src/generic/complex_matrices.h

```diff
@@ -377,6 +388,38 @@ namespace oomph
     void multiply_transpose(const Vector<std::complex<double>>& x,
                             Vector<std::complex<double>>& soln);
 
+    // \short Add the matrix to matrix_in (CRDoubleMatrix) and return result:
+    // result_matrix = A + matrix_in
+    void add(const CRDoubleMatrix& matrix_in,
+             CRComplexMatrix& result_matrix) const;
+
+    // \short Add the matrix to matrix_in (CRComplexMatrix) and return result:
+    // result_matrix = A + matrix_in
+    void add(const CRComplexMatrix& matrix_in,
+             CRComplexMatrix& result_matrix) const;
+
+    // \short Multiply the matrix by matrix_in (CRComplexMatrix) and returns
+    // result: result=A matrix_in
+    void multiply(const CRComplexMatrix& matrix_in,
+                  CRComplexMatrix& result) const;
+
+    /// \short Access function to Serial_matrix_matrix_multiply_method, the flag
+    /// which determines the matrix matrix multiplication method used for serial
+    /// matrices.
+    unsigned& serial_matrix_matrix_multiply_method()
+    {
+      return Serial_matrix_matrix_multiply_method;
+    }
+
+    /// \short Read only access function (const version) to
+    /// Serial_matrix_matrix_multiply_method, the flag
+    /// which determines the matrix matrix multiplication method used for serial
+    /// matrices.
+    const unsigned& serial_matrix_matrix_multiply_method() const
```

Comment:
This is probably related to my earlier query about "why do use a reference". Isn't the intended usage simply to get a read-only access to that unsigned? If so, 

       unsigned serial_matrix_matrix_multiply_method() const

is much neater (and is what we use everywhere else). It's then the direct (const) counterpart of the version which gives you a non-constant access,

        unsigned& serial_matrix_matrix_multiply_method()

Reply:
We need read and write access. First to check the default method is set correctly (to 2) and then to change the method to test the alternatives. I could use the access function by itself as:
```
matrix_square.serial_matrix_matrix_multiply_method() = 3;
```
etc. But I don't think assigning a function's output is very clear, though I think this is what we use elsewhere (correct me if I am wrong.)

Reply:
A method of manipulating the internal value is needed to change the multiplication approach. One approach, as Paul mentions is to assign to directly to the returned value. I dislike it because it's less readable than

```bash
method = 3;
```

which you can do by using a reference variable (the approach I recommended). If you wanted to do things "properly" (the answer to that of course should always be "yes"), you would create an `enum class`. 

**Note:** an `enum class` is slightly different to a plain enum; it is said to be "type-safe" because it avoids clashes with values from other enums (that evaluate to the same integer). See [here](https://stackoverflow.com/questions/18335861/why-is-enum-class-preferred-over-plain-enum) for some information. An important distinction is that you can't use the enum values as actual `int`s without an explicit cast. However, given that we probably only want to use the value in the following ways

```bash
// Assign
matrix_square.serial_matrix_matrix_multiply_method() = SerialMatrixMultiplyMethod::Fastest_approach;
matrix_square.serial_matrix_matrix_multiply_method() = SerialMatrixMultiplyMethod::Lowest_memory_usage_approach;

// Internal check
if (Serial_matrix_matrix_multiply_method == SerialMatrixMultiplyMethod:: Fastest_approach)
{
  // ...do the fast shit
}
```

I imagine it will be okay. It's worth checking though. (I.e. are there many cases where the explicit integer value is required?)

P.S. Despite the long reply, I'm too lazy to type out the entire solution, so consider it "left to the reader".

Reply:
I'd overlooked (what looks like a) magic number problem which we generally solve with an enum inside the class that uses it (which provides context and encapsulation). 

   Re working directly with references: This

```
unsigned& method = matrix_square.serial_matrix_matrix_multiply_method();
method = SerialMatrixMultiplyMethod::Fastest_approach;
method = SerialMatrixMultiplyMethod::Lowest_memory_usage_approach;
```
seems extremely odd to me.It's basically like working with a pointer (which is sort of what references)are, but the variable name doesn't make that clear. If it was a pointer, we'd have

```
unsigned* method_pt = matrix_square.serial_matrix_matrix_multiply_method_pt();
*method_pt = SerialMatrixMultiplyMethod::Fastest_approach;
*method_pt = SerialMatrixMultiplyMethod::Lowest_memory_usage_approach;
```
which makes the dereferencing explicit. (Note that I'm NOT advocating this approach here).

    Anyway, if a class has a member, `unsigned N`, we generally provide two access functions

```
/// Read access to N
unsigned n()  const {return N;}
```

and

```
/// Read/write access to N
unsigned&  n() {return N;}
```
 and we'd only set the value with

```
bla_class.n()=MagicNumberContainingNamespace::N1;
```
where the compiler then picks the read/write member function.

Reply:
@MatthiasHeilManchester, just in case you missed the comment, any opinions on this?